sdb/commands/array.py

```diff
@@ -0,0 +1,133 @@
+            (char)115
+            (char)112
+            (char)97
+            ... <cropped 95 other elements> ...
+
+        The command also works with pointer arrays as long as the
+        number of elements is specified:
+
+            sdb> spa | member spa_zio_taskq[0][0].stqs_taskq
+            *(taskq_t **)0xffff90cc154dcdc8 = 0xffff90cc127ea600
+
+            sdb> spa | member spa_zio_taskq[0][0].stqs_taskq | array 2
+            *(taskq_t *)0xffff90cc127ea600 = {
+                ... <cropped> ...
+            }
+            *(taskq_t *)0xffff90cc181bfb00 = {
+                ... <cropped> ...
+            }
+
+        Until a better way/notation is implemented on SDB the only
+        way to ask for specific subranges of elements is through
+        the addition of `head` and `tail` commands in the pipeline.
+    """
+
+    names = ["array"]
+
+    @classmethod
+    def _init_parser(cls, name: str) -> argparse.ArgumentParser:
+        parser = super()._init_parser(name)
+        parser.add_argument("nelems",
+                            type=int,
+                            nargs="?",
+                            help="number of elements in array to walk")
+        return parser
+
+    def _call(self, objs: Iterable[drgn.Object]) -> Iterable[drgn.Object]:
+        for obj in objs:
+            nelems = 0
+
+            obj_type = sdb.type_canonicalize(obj.type_)
+            if obj_type.kind == drgn.TypeKind.ARRAY:
+                array_elems = len(obj)
+                if self.args.nelems is not None:
+                    nelems = self.args.nelems
+                    if nelems > array_elems:
+                        print(
+                            f"warning: exceeding array size {array_elems} - requested {nelems}"
+                        )
+                else:
+                    nelems = array_elems
+
+            elif obj_type.kind == drgn.TypeKind.POINTER:
+                if self.args.nelems is None:
+                    err_msg = f"'{obj.type_.type_name()}' is a pointer - "
+                    err_msg += "please specify the number of elements"
+                    raise sdb.CommandError(self.name, err_msg)
+                nelems = self.args.nelems
+
+                if not obj_type.type.is_complete():
+                    err_msg = "can't walk pointer array of incomplete type "
+                    err_msg += f"'{obj_type.type.type_name()}'"
+                    raise sdb.CommandError(self.name, err_msg)
+
+                #
+                # We attempt to read from the address of the
+                # pointer array first to ensure that it is
+                # valid before attempting to dereference it
+                # later
+                #
+                try:
+                    obj[0].read_()
+                except drgn.FaultError:
+                    raise sdb.CommandError(
+                        self.name,
+                        f"invalid memory access at address {hex(obj.value_())}")
```

Comment:
I wonder if we should be catching this at a higher level, in the sdb infrastructure.  You could have other kinds of memory errors, like going off the end of the array into unmapped memory.